stylua-vscode/src/download.ts

```diff
@@ -0,0 +1,154 @@
+import * as vscode from "vscode";
+import * as unzip from "unzipper";
+import * as util from "./util";
+import fetch from "node-fetch";
+import { createWriteStream } from "fs";
+import { executeStylua } from "./stylua";
+import { GitHub, GitHubRelease } from "./github";
+
+export class StyluaDownloader {
+  constructor(
+    private readonly storageDirectory: vscode.Uri,
+    private readonly github: GitHub
+  ) {}
+
+  public async ensureStyluaExists(): Promise<string | undefined> {
+    const path = await this.getStyluaPath();
+
+    if (path === undefined) {
+      await vscode.workspace.fs.createDirectory(this.storageDirectory);
+      await this.downloadStyLuaVisual();
+      return await this.getStyluaPath();
+    } else {
+      if (!(await util.fileExists(path))) {
+        vscode.window.showErrorMessage(
+          `The path given for StyLua (${path}) does not exist`
+        );
+        return;
+      }
+
+      try {
+        const currentVersion = (
+          await executeStylua(path, ["--version"])
+        )?.trim();
+        const desiredVersion = util.getDesiredVersion();
+        const release = await this.github.getRelease(desiredVersion);
+        if (
+          currentVersion !==
+          `stylua ${
+            release.tagName.startsWith("v")
+              ? release.tagName.substr(1)
+              : release.tagName
+          }`
+        ) {
+          this.openUpdatePrompt(release);
+        }
+      } catch (err) {
+        vscode.window.showWarningMessage(
+          `Error checking the selected StyLua version, falling back to the currently installed version:\n${err}`
+        );
+        if (!this.github.authenticated) {
+          const option = await vscode.window.showInformationMessage(
+            "Authenticating with GitHub can fix rate limits.",
+            "Authenticate with GitHub"
+          );
+          switch (option) {
+            case "Authenticate with GitHub":
+              if (await this.github.authenticate()) {
+                return this.ensureStyluaExists();
+              }
+          }
+        }
```

Comment:
Should we just have a single prompt here, rather than two? It can get quite cluttered imo.

Reply:
I can change it to what you like, but when I was testing it, jamming it all into one prompt looked quite ugly. Unfortunately VSC doesn't think that notifications deserve formatting, so it removes any newlines or attempt at formatting.

Let me know your thoughts.

Reply:
An alternative we could use is to make it modal which does get formatting, but that may be too intrusive.

Reply:
I think a modal is too intrusive here. I think it's fine as it is for now.

Should we bother checking the error for a 429/ratelimit error before showing this specific prompt? Not really sure what other errors could happen

Reply:
We could do that. My thinking was that if we're getting errors and making more requests there's a risk of rate limit anyway, but if you want we can hold of displaying that until we start hitting the rate limit.

Reply:
No I think you're right, this is fine.

Reply:
We can always listen for feedback on it. This is a relatively easy fix if we need to later.